thorlcr/master/thgraphmanager.cpp

```diff
@@ -60,10 +61,157 @@ class CJobManager : public CSimpleInterface, implements IJobManager, implements
     StringAttr          currentWuid;
     ILogMsgHandler *logHandler;
 
+    CJobMaster *getCurrentJob() { CriticalBlock b(jobCrit); return jobs.ordinality() ? &OLINK(jobs.item(0)) : NULL; }
     bool executeGraph(IConstWorkUnit &workunit, const char *graphName, const SocketEndpoint &agentEp);
     void addJob(CJobMaster &job) { CriticalBlock b(jobCrit); jobs.append(job); }
     void removeJob(CJobMaster &job) { CriticalBlock b(jobCrit); jobs.zap(job); }
 
+    class CThorDebugListener : public CSimpleInterfaceOf<IInterface>, implements IThreaded
+    {
+    protected:
+        CThreaded threaded;
+        unsigned port;
+        Owned<ISocket> sock;
+        CJobManager &mgr;
+    private:
+        volatile bool running;
+    public:
+        CThorDebugListener(CJobManager &_mgr) : threaded("CThorDebugListener", this), mgr(_mgr)
+        {
+            port = globals->getPropInt("DebugPort", 9877);
+            running = true;
+            threaded.start();
+        }
+        ~CThorDebugListener()
+        {
+            running = false;
+            if (sock)
+                sock->cancel_accept();
+            threaded.join();
+        }
+        virtual unsigned getPort() { return port; }
+
+        virtual void processDebugCommand(CSafeSocket &ssock, StringBuffer &rawText)
+        {
+            Owned<IPropertyTree> queryXml;
+            try
+            {
+                queryXml.setown(createPTreeFromXMLString(rawText.str(), ipt_caseInsensitive, (PTreeReaderOptions)(ptr_ignoreWhiteSpace|ptr_ignoreNameSpaces)));
+            }
+            catch (IException *E)
+            {
+                StringBuffer s;
+                DBGLOG("ERROR: Invalid XML received from %s:%s", E->errorMessage(s).str(), rawText.str());
+                throw;
+            }
+
+            Linked<CJobMaster> job = mgr.getCurrentJob();
+            if (!job)
+                throw MakeStringException(5300, "Command not available when no graph active");//TODO:Update error code
+
+            const char *command = queryXml->queryName();
+            const char *edgeId = queryXml->queryProp("@edgeId");
+            const char *graphId = job->queryGraphName();
+
+            if (!command || !edgeId || !graphId) throw MakeStringException(5300, "Invalid debug command");
+
+            enum msgids cmdcode;
+            if (strncmp(command,"print", 5) == 0)
+                cmdcode = DebugRequest;
+            else if (strncmp(command,"quit", 4) == 0)
+                cmdcode = GraphAbort;
```

Comment:
It (IDE) should know what's supported, i.e. that it's targeting Thor and not enable functions that are not supported.
Or perhaps (in some future version) it could query for list of commands that are supported.

Quitting like this may cause unclean abort sequence (I'm not sure what will happen) - normally a job is aborted either when there's a failure and exceptions trickle up/around, or by CJobBase::abort() call, which e.g. can happen by user request via eclwatch workunit.
If you look for '"ABORT detected from user" in thgraphmaster.cpp, you can see how it handles it.
The debugger could handle "quit" in same way (that would cleaner), i.e. call job->fireException with TE_WorkUnitAborting.